### ISSUE TITLE
Offline Parsing results inconsistent with Online Parsing results

### ISSUE THREAD

**Issue by klaki892** [RickLeinecker/SeeShells#240]
_(Initially reported by @AlekStoyanov )_

Currently when SeeShells performs an offline parse on a `NTUSER.dat` file, the resulting shellbags are significantly less than what would be found if SeeShells performed online parsing on the same user account in which the `NTUSER.dat` file was produced from. 

This inconsistency reveals a flaw in the assumed programs functionality, that the same findings can be found both on a live machine and a postmortem machine. 
The program's inconsistent results are in direct contrast to our current understanding of shellbags which are:
1. The Shellbags are stored on a user level, in the user's respective hives.
2. Shellbag data is not ephemeral. It is recorded and stored and can be retrieved without needing the Windows OS running.  

Steps to reproduce:
1. Perform an Online parse on a machine
2. With the results limit it to a specific user and record the items found. (i.e. exporting results)
3. Perform an offline parse on the `NTUSER.dat` located in the specific users directory.
4. With the results, notice the disparity between the shellbag items found between both data sets.

Acceptance Criteria:
- Determine and document why the inconsistency currently exists 
    - (If Possible) resolve the issue causing the discrepancy between offline and online parsing results

**Comment by klaki892** [comment]
From initial research and observation: It appears that we currently are **not** looking in all of the right directories for shellbag information.

I've noticed on my personal machine and on a virtual machine that most of our shellbag results for _Online Parsing_ come from a separate user key underneath the `HKEY_USERS` in a live registry.

This is a filter on my user account (Notice the length of the scrollbar for comparison between the photos) 
![image](https://user-images.githubusercontent.com/17878901/77237094-5f3a9880-6b9b-11ea-8299-426bbdd3d7c4.png)

And this is a filter on the my user account, but the SID has `_Classes` attached to it:
![image](https://user-images.githubusercontent.com/17878901/77237119-b6406d80-6b9b-11ea-8bb8-a9973b82b1cd.png)


Both "Users" in the filter are the only results that appear, which is odd considering we assume that one user key in the HKEY_USERS corresponds to the respective NTUSER/UsrClass files for a user. 

![image](https://user-images.githubusercontent.com/17878901/77237129-cbb59780-6b9b-11ea-9805-54c687fcad24.png)

As the results show from the length of the scrollbar as a rough estimate for how many shellbags are in each user filter, it would appear _most of our shellbags for online parsing are coming from the user with `_Classes` in the identifier._ 

Researching why _Classes users exist lead me to the following information about HKEY_USERS:

> The HKEY_USERS subtree contains all actively loaded user profiles. HKEY_USERS has at least three keys.
> 
> ...
**> A subkey named for the Security Identifier (SID) of the current local user with the _Classes suffix. This subkey contains the current user's Classes. The data in HKEY_USERS\<SID>\_Classes is also contained in HKEY_CLASSES_ROOT.**
> In Windows 2000 the default user profile is not stored in the registry. It is stored on the system drive in \Documents and Settings\Default User [WINNT]\Ntuser.dat.
([Source](http://systemmanager.ru/win2k_regestry.en/69674.htm))

Going down the rabbit hole, this is some relevant information about the HKEY_CLASSES_ROOT, and the location of which this Registry store is generated:

> HKEY_CLASSES_ROOT is a global hive, which means it can contain information that applies to all the users on the computer and is viewable by every user. This is in contrast to some hives that have information that applies only to the currently signed-in user.
> 
> However, because the HKEY_CLASSES_ROOT hive is actually combined data found in both the HKEY_LOCAL_MACHINE hive (HKEY_LOCAL_MACHINE\Software\Classes) **and the HKEY_CURRENT_USER hive (HKEY_CURRENT_USER\Software\Classes), it also contains user-specific information as well.** Even though that's the case, the HKEY_CLASSES_ROOT is still able to be browsed by any and all users.
([Source](https://www.lifewire.com/hkey-classes-root-2625899))

It would appear as though the missing shellbags in offline parsing is a result of our current OS Configuration  **Not looking** underneath `Software\Classes` in a user's registry hive

However I'm hesitant to say these are conclusive results because both the online and offline parsing methods look at the **same OS configuration list** for where to find shellbags. And through my own registry browsing i was unable to find a path like `Software\\Microsoft\\Windows\\Shell\\Bags` underneath the _Classes user Key in my registry:
![image](https://user-images.githubusercontent.com/17878901/77237313-8a25ec00-6b9d-11ea-8743-126964322aa8.png)

From this initial research my follow up questions would be:
1. Why does the `_classes` user under HKEY_USERS contain many shellbags? 
2. What is the full filepath of the shellbags that we are retrieving from `SID_Classes` when we do an online parse.
    - (At the time of writing our program does not reveal the full registry path of the shellbag in which the `ShellItem` was identified, which might severely hinder us in finding the answers to this overall issue)

**Comment by szczura** [comment]
Partially addressed by #311

**Comment by klaki892** [comment]
Follow up to previous research:

**TL;DR: Our Usual testing only had NTUser.dat, we never tested with UsrClass.dat. reading from a live registry had both contained under two keys in `HKEY_USERS` the one with the SID (ntuser.dat) and the one with SID_Classes (usrclass.dat)**

>Why does the _classes user under HKEY_USERS contain many shellbags?
Simply put by the research Vincent Lo, the disparity in the amounts of shellbags between the files can be attributed to the follow explanation:

> ... the NTUSER.DAT stores the ShellBag information for
> the Desktop, Windows network folders, remote machines and remote folders. The
> UsrClass.dat stores the ShellBag information for the Desktop, ZIP files, remote folders,
> local folders, Windows special folders and virtual folders. 
>  - [Vincent Lo,  Windows ShellBag Forensics in Depth](https://www.sans.org/reading-room/whitepapers/forensics/windows-shellbag-forensics-in-depth-34545)

> What is the full filepath of the shellbags that we are retrieving from SID_Classes when we do an online parse.
> (At the time of writing our program does not reveal the full registry path of the shellbag in which the ShellItem was identified, which might severely hinder us in finding the answers to this overall issue)
>
This turned out to be a trick question due to lack of understanding and not enough looking into the Registry side of shellbags and focusing on the values of the bags themselves.  With the completion of #277 we could see the registry paths for which the shellbags came from, and it was upon this inspection of live parsing that I rediscovered something [the shellbag community was already heavily aware of](https://www.dfrws.org/sites/default/files/session-files/paper-using_shellbag_information_to_reconstruct_user_activities.pdf) (page 2-3), the separate locations of metadata about a shellbag and the actual values of the shellbag. Due to BagMRU's nature of hierarchically storing the folder structure observed, you could see the metadata (path to a file) and the NodeSlot, the location of the actual shell data for the item. Although iterated over, our traversal over the keys simply converted from BagMRU's NodeSlot path to the respective location in Shell\Bags and pulled their respective values. Never storing the metadata surrounding the path it took to obtain the NodeSlot. So In Summary
-BagsMRU registry keys contain (or the metadata surrounding them) Shell\Bags gives us the actual data of the shellbag. Both are needed together to provide relevant information in a grand picture scheme. With the creation of #246 this can be seen visually :
![image](https://user-images.githubusercontent.com/17878901/79036917-027b3e00-7b9b-11ea-97d2-f02671470336.png)
